kernel/include/kernel_arch.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2016 Wind River Systems, Inc.
+ * Copyright (c) 2019 Stephanos Ioannidis <root@stephanos.io>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_KERNEL_INCLUDE_KERNEL_ARCH_H_
+#define ZEPHYR_KERNEL_INCLUDE_KERNEL_ARCH_H_
+
+#if !defined(_ASMLANGUAGE)
+
+#include <string.h>
+#include <sys/arch_interface.h>
+
+#include <kernel_structs.h>
+#include <kernel_arch_func.h>
+
+#ifdef CONFIG_USE_SWITCH
+/* This is a arch function traditionally, but when the switch-based
+ * z_swap() is in use it's a simple inline provided by the kernel.
+ */
+static ALWAYS_INLINE void
+z_arch_thread_return_value_set(struct k_thread *thread, unsigned int value)
```

Comment:
duplicate definition.
this is already defined in include/sys/arch_interface.h

Reply:
well, the one in include/sys/arch_interface.h does not include implementation of it.

Reply:
that's because it gets defined by the arch code, typically in arch/*/include/kernel_arch_func.h

everything defined in include/sys/arch_interface.h has its implementation in arch-specific code, there's no general implementation for them

Reply:
> everything defined in include/sys/arch_interface.h has its implementation in arch-specific code, there's no general implementation for them

Yes, but the point of this code is to provide implementation of it when CONFIG_USE_SWITCH is enabled, in which case arch-specific code should not provide an implementation (at least, according to the comment here):

https://github.com/zephyrproject-rtos/zephyr/pull/20047/files/426c6cad08f7869f734bb3a8c3d681aa0c7b2e2a#diff-7d60221bd6a21c05f5f30c3d0ff07e7f

Reply:
Note that this code was simply moved out of kernel_structs.h to kernel_arch.h (as it certainly does not belong in kernel_structs.h).

Reply:
Just to be sure, I further validate this as follows:

1. In k_thread, swap_retval field is added when CONFIG_USE_SWITCH is enabled.
https://github.com/zephyrproject-rtos/zephyr/blob/9784f800a6866c636af0d00a9de0f5b92000dc31/include/kernel.h#L574-L584

2. do_swap uses k_thread swap_retval as the thread return value.
https://github.com/zephyrproject-rtos/zephyr/blob/9784f800a6866c636af0d00a9de0f5b92000dc31/kernel/include/kswap.h#L69-L93

Reply:
the way to solve this is to add another layer of indirection.

define a z_thread_return_value_set(), which either calls into the arch API or sets the thread member depending on whether SWITCH is enabled

Reply:
> the way to solve this is to add another layer of indirection.

Yes, that would be the right way to do it. I am somewhat reluctant to introduce that change in this PR though due to the possibility of bringing on yet another another dependency hell, which this PR already has more than enough of.

Maybe it could be addressed in another upcoming arch interface-related PR.